### PR TITLE
Make `Add-Type` usable in application that hosts PowerShell

### LINKED ISSUE #9587: 'Add-Type' doesn't work in applications that host powershell
`Add-Type` depends on the reference assemblies shipped with `pwsh`. The reference assemblies are not bundled with PowerShell SDK NuGet packages, so `Add-Type` doesn't work for applications that host PowerShell.

### PR DESCRIPTION
**Pull request by daxian-dbw**
# PR Summary

Fix #9587 
- Make the NuGet package `Microsoft.PowerShell.SDK` ship the `ref` folder, which will be copied to the right place for `dotnet build` and `dotnet publish`.
- Make change in `Add-Type` to look for the `ref` folder at the folder where `EntryAssembly` is located. In case of `pwsh`, it's the folder that contains `pwsh.dll`.
- Add a hosting test to make sure `Add-Type` works for an application that hosts powershell.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs [hunk 1]
```diff
@@ -599,7 +599,7 @@ protected override void EndProcessing()
         // We now ship .Net Core's reference assemblies with PowerShell, so that Add-Type can work
```
Comment (PoshChan) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs]: @daxian-dbw, your last commit had 4 failures in `PowerShell-CI-windows`
Tests Import-PSSession cmdlet works with types unavailable on the client.Verifies client-side unavailable enum is correctly handled
```powershell
Expected 2, but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Implicit.Remoting.Tests.ps1: line 130
130:                 (foo -x "Value2") | Should -Be 2
```
Import-PSSession with FormatAndTypes.Implicit remoting works even when types.ps1xml is missing on the client.Serialization works for top-level properties
```powershell
Expected exactly 'root', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Implicit.Remoting.Tests.ps1: line 702
702:                 $x.text | Should -BeExactly "root"
```
Import-PSSession with FormatAndTypes.Implicit remoting works even when types.ps1xml is missing on the client.Serialization settings works for deep properties
```powershell
Expected exactly 'Grandson', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Implicit.Remoting.Tests.ps1: line 707
707:                 $x.Son.Grandson.text | Should -BeExactly "Grandson"
```
Import-PSSession with FormatAndTypes.Implicit remoting works even when types.ps1xml is missing on the client.Serialization settings are preserved even if types.ps1xml is missing on the client
```powershell
Expected exactly 'Grandson', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Implicit.Remoting.Tests.ps1: line 712
712:                 $y | Should -BeExactly "Grandson"
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs [hunk 2]
```diff
@@ -599,7 +599,7 @@ protected override void EndProcessing()
...
         // in a predictable way and won't be broken when we move to newer version of .NET Core.
         // The reference assemblies are located at '$PSHOME\ref'.
-        private static readonly string s_netcoreAppRefFolder = PathType.Combine(PathType.GetDirectoryName(typeof(PSObject).Assembly.Location), "ref");
+        private static readonly string s_netcoreAppRefFolder = PathType.Combine(PathType.GetDirectoryName(Assembly.GetEntryAssembly().Location), "ref");
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs]: Could you please clarify why we need the change? Should we update comment above too?

Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs]: Could you please clarify why we need the change? PSObject is in SMA, Add-Type is in Microsoft.PowerShell.Utility.dll.
 Should we update comment above too?

File: test/hosting/test_HostingBasic.cs [hunk 3]
```diff
@@ -139,5 +139,37 @@ public static void TestCommandFromNative()
                 File.Delete(target);
             }
         }
+
+        [Fact]
+        public static void TestAddType()
```
Comment (iSazonov) [inline test/hosting/test_HostingBasic.cs]:
```suggestion
        public static void TestAddTypeCmdletInHostScenario()
```

File: test/hosting/test_HostingBasic.cs [hunk 4]
```diff
@@ -139,5 +139,37 @@ public static void TestCommandFromNative()
...
+                    public string Path;
+                }
+            ";
+
+            using (System.Management.Automation.PowerShell ps = System.Management.Automation.PowerShell.Create())
+            {
+                ps.AddCommand("Add-Type").AddParameter("TypeDefinition", code).Invoke();
+                ps.Commands.Clear();
+
+                var results = ps.AddScript("[Foo]::new('Joe', 'Unknown')").Invoke();
```
Comment (iSazonov) [inline test/hosting/test_HostingBasic.cs]: Should we be more strong in tests?
```suggestion
                Collection<PSObject> results = ps.AddScript("[Foo]::new('Joe', 'Unknown')").Invoke();
```

File: test/hosting/test_HostingBasic.cs [hunk 5]
```diff
@@ -139,5 +139,37 @@ public static void TestCommandFromNative()
...
+                Assert.Single(results);
+
+                dynamic foo = results[0];
```
Comment (iSazonov) [inline test/hosting/test_HostingBasic.cs]: The same question.

File: test/hosting/test_HostingBasic.cs (hunk 4 above)
Comment (daxian-dbw) [reply test/hosting/test_HostingBasic.cs]: I think it's fine to use `var` here. The purpose of the tests here is to make sure basic things work in hosting scenario, not specifically testing the API itself.
Also, changing signature of the API is too much a breaking change that will probably never be accepted :)

File: test/hosting/test_HostingBasic.cs (hunk 5 above)
Comment (daxian-dbw) [reply test/hosting/test_HostingBasic.cs]: Same here. Also, I have to use `dynamic` to be able to access the properties of the underlying .NET type directly.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs (hunk 2 above)
Comment (daxian-dbw) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs]: I have updated the comments. Hopefully it's clear enough about the intention

File: test/hosting/test_HostingBasic.cs (hunk 3 above)
Comment (daxian-dbw) [reply test/hosting/test_HostingBasic.cs]: Name updated.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs [hunk 6]
```diff
@@ -598,8 +598,19 @@ protected override void EndProcessing()
...
+        // The reference assemblies are located at '$PSHOME\ref' for pwsh.
+        //
+        // For applications that host PowerShell, the 'ref' folder will be deployed to the 'publish'
+        // folder, not where 'System.Management.Automation.dll' is located. So here we should use
+        // the entry assembly's location to construct the path to the 'ref' folder.
+        // For pwsh, the entry assembly is 'pwsh.dll', so the entry assembly's location is still
+        // $PSHOME.
+        // However, 'Assembly.GetEntryAssembly()' returns null when the managed code is called from
+        // unmanaged code (PowerShell WSMan remoting scenario), so in that case, we continue to use
+        // the location of 'System.Management.Automation.dll'.
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs]: @daxian-dbw Thanks for the great comment! Taking in account that we have very few docs for SDK and hosting I think it makes sense to open new documentation issue for this.

**Review by iSazonov — APPROVED** [review]

**Review by TylerLeonhardt — APPROVED** [review]
Such a small change for big impact 😊 LGTM

File: test/hosting/test_HostingBasic.cs [hunk 7]
```diff
@@ -139,5 +139,37 @@ public static void TestCommandFromNative()
                 File.Delete(target);
             }
         }
+
+        [Fact]
+        public static void TestAddTypeCmdletInHostScenario()
```
Comment (PaulHigin) [inline test/hosting/test_HostingBasic.cs]: Please add a comment that describes the purpose of the test (e.g., 'Verifies reference assemblies in PowerShell hosted scenario').

**Review by PaulHigin — APPROVED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs (hunk 6 above)
Comment (daxian-dbw) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs]: As for documentation, I think we should handle the reference assemblies in a way that it just works, and users can use `Add-Type` without needing to consider how it works.
From that perspective, we don't need to surface this in the help docs. It should be internal implementation docs if we want to doc it.

File: test/hosting/test_HostingBasic.cs (hunk 7 above)
Comment (daxian-dbw) [reply test/hosting/test_HostingBasic.cs]: Comment added.

**Review by adityapatwardhan — APPROVED** [review]

**Comment by TylerLeonhardt** [comment]
@Francisco-Gamino fyi

**Pull request merged**